Converts Slime processor process to a Proximity Monitor

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
closes: #4384
This PR converts the current process of the slime processor to a Proximity Monitor as it is way faster and also means one object less thats processing also sets the parent of the slime processors processing_flags to none as it also called processing despite it not even having a process proc.

~~Only drawback is that if the slime processor is currently active and you put additional slimes down you will have to pick those up again and place them down or move them around so that it calls hasproximity again.~~ Now there is a fallback to hearers after it finishes to process the slimes so thats no longer a problem.
## Why It's Good For The Game

Ike wants this

## Changelog
:cl:
code: Rewrote the slime processor to be more performant
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
Converts Slime processor process to HasProximity

Converts the slime processor process to has proximity also sets the processors process flag to none as without that even the parent will call process

MNarath1 committed:
change order in if statement

changes the order of istype and stat == DEAD in if statement

MNarath1 committed:
deletes completly useless check if the slime is alive

MNarath1 committed:
lets stay consistant

utilbird commented:
> Only drawback is that if the slime processor is currently active and you put additional slimes down you will have to pick those up again and place them down or move them around so that it calls hasproximity again.

You could make it call it when the slime processor finishes processing instead

MNarath1 commented:
> 
> 
> > Only drawback is that if the slime processor is currently active and you put additional slimes down you will have to pick those up again and place them down or move them around so that it calls hasproximity again.
> 
> You could make it call it when the slime processor finishes processing instead

yeah trough the problem is due to how proximity monitor works i can't use that to recheck a not moved mob trough i guess i could make an exception with ohear again thats called after the processor finishes

MNarath1 committed:
use hearers as fallback in case new slimes got placed during processing

Just a fallback to hearers in case someone placed new slimes in that time

MNarath1 commented:
Now i know calling hearers() is not ideal but this gets only called once after the processor finishes to scoop up all thats left around instead of getting called almost every tick as a process. Also i am using hearers for this instead of ohearers as the latter one ignores the center.

MNarath1 committed:
reduces call amount of hearers even further

now this might be a bit overdoing trough it will reduce the call amouts of hearers to really only when actually needed

MNarath1 commented:
now the last commit might be a bit overdoing trough it reduces the amounts of needed calls of hearers even further now i know i could also save all the objects into a list and then force move them later on trough that would not account for if it gets moved out of range again thats why i still call hearers to check if those are still in the area.

MNarath1 committed:
oops

no dam reason to set the var every single loop

ike709 commented on code/modules/food_and_drinks/kitchen_machinery/processor.dm:
```diff
@@ -173,31 +176,27 @@
+/obj/machinery/processor/slime/interact(mob/user)
+	. = ..()
+	if(sbacklogged)
+		for(var/mob/living/simple_animal/slime/AM in hearers(1,src)) //fallback in case slimes got placed while processor was active
```
```suggestion
		for(var/mob/living/simple_animal/slime/AM in ohearers(1,src)) //fallback in case slimes got placed while processor was active
```

ike709 replied:
Also this would move all of the adjacent slimes, defeating the point of `processing`

MNarath1 replied:
reason of why i use hearers here instead of ohearers is cause hearers includes the center too i know from playing xenobio that many players usually place slimes directly ontop of the processor

MNarath1 replied:
Also this is simply there to act as an fallback if someone places down slimes during the time where it processes because the proximity monitor won't trigger for the slimes again except if someone moves them

MNarath1 replied:
after it finishes it simply then pulls in all the slimes adjacent to it yes trough that will only trigger after the interact proc of the parent is finished and at that time processing will be set to false again

MNarath1 replied:
its basically just a way to recheck for all the slimes that got put down while the processor was active afterwards because the proximity monitor can't do that so those won't just keep sitting there

MNarath1 committed:
clarifing a comment

This might have been a bit missleading this part of the interact proc does not trigger during processing but after its finished and only if the  Proximity Monitor actually detected a death slime during processing.

MNarath1 commented:
@ike709 I think i understand now how this missunderstanding came to be. The part of the interact proc where hearers gets called does not get called during processing but after processing finishes also it gets only called if the proximity monitor detects a death slime during processing. Reason for this to be hearers instead of ohearers is simply cause hearers includes the center unlike ohearers if i would use ohearers this  would also be inconsistant behavior as the proximity monitor also triggers on movement to  the center / teleporting slimes on top of the processor.

MNarath1 commented:
without this the slimes that got placed during processing will not be pulled in afterwards as the proximity monitor will not trigger unless the slimes get moved

MNarath1 committed:
replace hearers with ohearers (for now)

might make a follow up PR later on to change to something completly different but i guess its fine for now.

MNarath1 commented:
I guess using ohearers for now is fine i will trough try to change this later in a follow up PR trough for that i first need https://github.com/BeeStation/BeeStation-Hornet/pull/4454 to be merged cause what i want to change pretty much needs what this stuff changes so changing the proximity monitor now makes little sense. 
Basic idea is to modifiy the proximity monitor so it can be set to record when a object enters its range and also when it leaves so it actually keeps updated what objects are in proximity to the object. Trough like i said cause crossed and uncrossed get replaced with signals changing this now would be pretty dam useless so i wait.
